core/engine/src/value/inner/nan_boxed.rs

```diff
@@ -0,0 +1,864 @@
+//! This `[JsValue]` inner type is a NaN-boxed value, which is a 64-bits value
+//! that can represent any JavaScript value. If the integer is a non-NaN value,
+//! it will be stored as a 64-bits float. If it is a `f64::NAN` value, it will
+//! be stored as a quiet `NaN` value. Subnormal numbers are regular float.
+//!
+//! For any other type of values, the value will be stored as a 51-bits non-zero
+//! integer.
+//!
+//! In short, the memory layout of a NaN-boxed value is as follows:
+//!
+//! | Type of           | Bit Layout | Comment |
+//! |-------------------|------------|---------|
+//! | `+Infinity`       | `7FF0:0000:0000:0000`    | |
+//! | `-Infinity`       | `FFF0:0000:0000:0000`    | |
+//! | `NAN` (quiet)     | `7FF8:0000:0000:0000`    | |
+//! | `NAN` (signaling) | `FFF8:0000:0000:0000`    | |
+//! | `Undefined`       | `7FF4:0000:0000:0000`    | |
+//! | `Null`            | `7FF5:0000:0000:0000`    | |
+//! | `False`           | `7FF6:0000:0000:0000`    | |
+//! | `True`            | `7FF6:0000:0000:0001`    | |
+//! | `Integer32`       | `7FF7:0000:IIII:IIII`    | 32-bits integer. |
+//! | `BigInt`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 0` | 51-bits pointer. Assumes non-null pointer. |
+//! | `Object`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 1` | 51-bits pointer. |
+//! | `Symbol`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 2` | 51-bits pointer. |
+//! | `String`          | `7FF[8-F]:PPPP:PPPP:PPPP \| 3` | 51-bits pointer. |
+//! | `Float64`         | Any other values.        | |
+//!
+//! Pointers have the highest bit (in the `NaN` tag) set to 1, so they
+//! can represent any value from `0x8000_0000_0000` to `0xFFFF_FFFF_FFFF`.
+//! The last 2 bits of the pointer is used to store the type of the value.
```

Comment:
is this right or am I reading it wrong? Looks like the value type (for pointers) is stored from `7FF8... - 7FFF...` at the top end (when looking at the values above from line 22 to 25)

Reply:
That's correct. If we only had 1 pointer type it would be easier, but I used a combination of pointer tagging and NaN Boxing to cover all our cases.